spark/src/main/java/org/apache/iceberg/spark/source/StreamingOffset.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.io.StringWriter;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+
+/**
+ * An implementation of Spark Structured Streaming Offset, to track the current processed
+ * files of Iceberg table. This StreamingOffset is combined by:
+ *
+ * version: used to validate the version number of StreamingOffset when deserializing from json
+ * string.
+ * snapshot_id: used to record the current processed snapshot of Iceberg table.
+ * index: used to record the index of scan files in this snapshot.
+ * is_starting_snapshot: used to identify if this snapshot is a starting snapshot. For starting snapshot
+ * we should scan all the files, not just incrementally add files.
+ * last_index_of_snapshot: used to denote whether the current index is the last file of this snapshot,
```

Comment:
Like "isStartingSnapshot" I think we can replace this with a more meaningful name like "snapshot_fully_processed" since "last_index_of_snapshot" is an implementation detail meaning "snapshot_fully_processed" right?

Reply:
Sure, will change to a more meaningful name.